src/hotspot/share/nmt/vmatree.hpp

```diff
@@ -0,0 +1,375 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_NMT_VMATREE_HPP
+#define SHARE_NMT_VMATREE_HPP
+
+#include "memory/resourceArea.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "nmt/nmtTreap.hpp"
+#include "runtime/os.hpp"
+#include "utilities/growableArray.hpp"
+#include "nmt/nmtNativeCallStackStorage.hpp"
+
+
+class VMATree {
+  static int addr_cmp(size_t a, size_t b);
+public:
+  enum class InOut {
+    Reserved,
+    Committed,
+    Released
+  };
+
+  // Each node has some stack and a flag associated with it.
+  struct Metadata {
+    NativeCallStackStorage::StackIndex stack_idx;
+    MEMFLAGS flag;
+
+    Metadata()
+      : stack_idx(),
+        flag(mtNone) {
+    }
+    Metadata(NativeCallStackStorage::StackIndex stack_idx, MEMFLAGS flag)
+      : stack_idx(stack_idx),
+        flag(flag) {
+    }
+    static bool equals(const Metadata& a, const Metadata& b) {
+      return NativeCallStackStorage::StackIndex::equals(a.stack_idx, b.stack_idx) &&
+             a.flag == b.flag;
+    }
+  };
+
```

Comment:
Hi, just to be clear: We don't perform a tree search whenever `is_noop` is used, the `leqA_n` is also present in your draft. Still, I think you might be right that having the full state for both in and out is the best way to go, I'll try it out as a simplification.

Reply:
Ah, you are of course right.

I remember trying your approach first, only keeping one state per node (also the outgoing state). Because, like you, I thought that the redundant data are unnecessary. But I ran into problem with rare corner cases, so I switched to the in-out-model. 

Unfortunately I blank out when trying to remember what the issue actually was. Figures :)